### PR TITLE
Support renaming

### LINKED ISSUE #201: Tracking file rename
In Jupyter we have a [new service](https://github.com/jupyter-server/jupyter_server_fileid) that allows to track files when they are renamed. It currently implements most of the logic, but I was wondering if this is something that watchfiles could support.
I know [Notify](https://github.com/notify-rs/notify) doesn't support that feature, but maybe this could be a nice addition to watchfiles?

### PR DESCRIPTION
**Pull request by davidbrochart**
Closes #201.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -37,7 +40,7 @@ enum WatcherEnum {
 
 #[pyclass]
 struct RustNotify {
-    changes: Arc<Mutex<HashSet<(u8, String)>>>,
+    changes: Arc<Mutex<HashSet<(u8, String, String)>>>,
```
Comment (samuelcolvin) [inline src/lib.rs:43]:
```suggestion
    changes: Arc<Mutex<HashSet<(u8, Option<String>, String)>>>,
```

maybe? Then you can use `None` for the `src` for `added`, `moved` and `deleted`?

File: src/lib.rs [hunk 2]
```diff
@@ -23,10 +23,13 @@ create_exception!(
     "Internal or filesystem error."
 );
 
+const EMPTY_STRING: String = String::new();
+
 // these need to match `watchfiles/main.py::Change`
 const CHANGE_ADDED: u8 = 1;
 const CHANGE_MODIFIED: u8 = 2;
 const CHANGE_DELETED: u8 = 3;
+const CHANGE_MOVED: u8 = 4;
```
Comment (samuelcolvin) [inline src/lib.rs:32]: we should probably use an enum for these.

Either an enum with values, or:

```rs
enum Change {
  Added(String),
  Modified(String),
  Deleted(String),
  Moved(String, String),
}
```
Then implement `ToPyObject` on Change.

File: src/lib.rs [hunk 3]
```diff
@@ -120,23 +123,30 @@ impl RustNotify {
...
-                            // but no information about whether it's the src or dst path is available.
-                            // Hence we have to check if the file exists instead.
-                            if Path::new(&path).exists() {
-                                CHANGE_ADDED
-                            } else {
-                                CHANGE_DELETED
-                            }
+                        EventKind::Modify(ModifyKind::Name(RenameMode::Both)) => {
+                            let mut changes = changes_clone.lock().unwrap();
+                            changes.remove(&(CHANGE_DELETED, path.clone(), EMPTY_STRING));
```
Comment (samuelcolvin) [inline src/lib.rs:128]: Does this actually happen? If so, I guess we should remove `added` here too?

**Review by samuelcolvin — COMMENTED** [review]
otherwise looks like a good start.

File: src/lib.rs (hunk 3 above)
Comment (davidbrochart) [reply src/lib.rs:128]: It happens when a file is moved out of the directory. In this case we want to see it as deleted, which is handled in the `RenameMode::From` event. But if the file is moved in the same directory, we want to see it as moved, so we need to remove the `CHANGE_DELETED`. I didn't find a better way to do it.
When a file is moved in the directory from outside, we want to see it as added, this works fine already.

File: src/lib.rs (hunk 2 above)
Comment (davidbrochart) [reply src/lib.rs:32]: Thanks, I will do that later. This is the first time I write Rust, so it's all new for me :smile:

File: src/lib.rs (hunk 1 above)
Comment (davidbrochart) [reply src/lib.rs:43]: Yes I tried that before, and run into issues that I didn't really understand.

File: src/lib.rs (hunk 3 above)
Comment (samuelcolvin) [reply src/lib.rs:128]: makes sense.

File: src/lib.rs (hunk 2 above)
Comment (samuelcolvin) [reply src/lib.rs:32]: I see, then I'm impressed.

Let me know if you have more questions, or if you want some help with this.

Lots of examples of `ToPyObject` in pydantic-core.

Also github copolit is very useful when you're new to rust (or indeed not new to it), it makes some very helpful suggestions.

**Comment by davidbrochart** [comment]
It's getting difficult to handle the macOS case, where renaming a file doesn't provide the src/dst paths. I'm starting to think that the [other option](https://github.com/samuelcolvin/watchfiles/issues/201#issuecomment-1301889791) you mentioned would be better: "process the simplified stream of events that we produce now and try to identify renaming".

**Comment by davidbrochart** [comment]
Yes, I'll open another PR where we post-process events in Python, but we keep track in Rust of "rename from X" and "rename to Y" events, and I'll get rid of the "rename from X to Y" event. I think it will make it easier to handle these special cases like the one on macOS.

**Comment by samuelcolvin** [comment]
Are you really sure macos events don't include src and dst? Where did you find that?

I know events aren't very good in macos (compared to linux), but I'm surprised by that. I'll check on my mac when I get home.

Sorry you didn't have more luck with this.

**Comment by samuelcolvin** [comment]
One option could be to collect `Rename::To` events and `Rename:From` events, then only try to fit them together to form rename changes, rather than look at all changes. The problem with this is we loose rename events for the polling watcher which would be sad given I think quite a few people are using the poll watcher.

See [`RenameMode`](https://docs.rs/notify/latest/notify/event/enum.RenameMode.html#)

**Comment by davidbrochart** [comment]
> Are you really sure macos events don't include src and dst? Where did you find that?

From [these comments](https://github.com/samuelcolvin/watchfiles/blob/a045bde477d3b50931b26bc58ffac1f58ff71acc/src/lib.rs#L127-L129), and the [tests](https://github.com/samuelcolvin/watchfiles/actions/runs/3394033562/jobs/5642085731) confirm that.

> Sorry you didn't have more luck with this.

No worries, maybe this is how my journey to Rust begins :smile: 

> One option could be to collect `Rename::To` events and `Rename:From` events, then only try to fit them together to form rename changes, rather than look at all changes.

Yes, that's what I meant in [my comment](https://github.com/samuelcolvin/watchfiles/pull/202#issuecomment-1303781270).

**Comment by davidbrochart** [comment]
Thinking about it, it is impossible to only use `Rename:From` and `Rename:To` to infer that a file was renamed from one path to another, because there could be multiple files renamed in a single change set, right?

**Comment by samuelcolvin** [comment]
> Thinking about it, it is impossible to only use `Rename:From` and `Rename:To` to infer that a file was renamed from one path to another, because there could be multiple files renamed in a single change set, right?

Well that's true, there can be hundreds of files changed in a single set (e.g. when you change branch in git), and we have no file hash to de-duplicate with.

The only options are:
* match "move froms" to "move tos" via file name for all files (e.g. added, deleted, `Rename:From`, `Rename:To`) and accept we'll get lots of false positives - this would have the added advantage of working for the polling watcher as well as other watchers
* match "move froms" to "move tos" via file name for `Rename:From` & `Rename:To` events only and therefore - with this we could get some false matches, but we wouldn't get lots of false positives, but I guess it wouldn't work for the polling watcher

I think we cal also get date modified and maybe even size on a file from the OS without reading the file - if we had them, we could do a much better job of matching files. WDYT?

**Comment by davidbrochart** [comment]
> I think we cal also get date modified and maybe even size on a file from the OS without reading the file - if we had them, we could do a much better job of matching files. WDYT?

Yes, that's what is done in [jupyter_server_fileid](https://github.com/jupyter-server/jupyter_server_fileid/blob/328d893ff2323f20925e036e57eb62f302fa94e2/jupyter_server_fileid/manager.py#L498-L513), but I thought we wouldn't have to do that if we used Notify. In the end, it seems that move events only work on Linux?

**Comment by samuelcolvin** [comment]
It sounds like we have to go the whole hot:
* detect move events with from and to on linux and mark them as a "move"
* collect "move from" and "move to" events and put them in a set to be matched at the end of a step
* optionally do the same for all events if we have the polling watcher

I think to avoid this being a breaking change, and to avoid extra processing when this isn't required, we should put this logic behind a flag/argument parameter.

Do you agree with that?

I haven't reviewed this PR recently, do you want me to review it or should we work on this separeately?

Also, are you willing to have a crack at this, or do you want me to work on it? Would be great if you could, but I understand it's quite a complex feature to start your rust career. I'm happy to work on it, but won't be for a few weeks at least.

**Comment by davidbrochart** [comment]
> Do you agree with that?

This seems like a good plan.

> I haven't reviewed this PR recently, do you want me to review it or should we work on this separeately?

I reverted my last attempts and got back to the start, where I just assumed that `Rename:Both` was working on all platforms. But I think it's better to open a new PR.

> Also, are you willing to have a crack at this, or do you want me to work on it?

I'm not sure, it seems that it will take me a lot of time, and you will probably do a much better work.

**Comment by samuelcolvin** [comment]
okay, makes sense. I'll try and work on it when I get time.

Thanks so much for investigating this, it's made my work much easier.

I'll leave this open to remind me about it.

**Comment by samuelcolvin** [comment]
closing this as I haven't needed it so haven't work on it, happy to review a new PR if someone wants to work on it.

**Pull request closed without merging**